In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
def fun_1(x):   #it is a function as returning list of squares

  A=[i*i for i in x]  #creating list
  return A

print(fun_1)
list=[1,2,3]
l=fun_1(list)
print('returned output is list',l)
    
def fun_gen(x):   #it is a generator  as yielding values 
  for i in x:
    yield (i*i)
a=[4,5,6]
l=fun_gen(a) #here l is a generaotor object
print('returned output is ',l)
print('values generated :')
for val in l:  #retrieving  values 
  print(val)

def fun_generator(professionals):
    for i in xrange(professionals):
        person = {
                    'id': i,
                    'name': random.choice(names),
                    'profession': random.choice(profession)
                }
        yield person
def function_list(professional):
    result = []
    for i in range(professional):
        person = {
                    'id': i,
                    'name': random.choice(names),
                    'profession': random.choice(profession)
                }
        result.append(person)
    return result


<function fun_1 at 0x00000239389140D0>
returned output is list [1, 4, 9]
returned output is  <generator object fun_gen at 0x00000239389164A0>
values generated :
16
25
36


**Use of generators :**
Generator functions allow you to declare a function that behaves like an iterator. Iterators don’t compute the value of each item when instantiated. They only compute it when you ask for it. This is known as lazy evaluation.Lazy evaluation is useful when you have a very large data set to compute. It allows you to start using the data immediately, while the whole data set is being computed.


In [4]:
myfirstlist = [1, 10,9,0]
listmy = [k**2 for k in myfirstlist]   #creating list
generator = (k**2 for k in myfirstlist) # () : paranethesis creating generator instead of list
print(listmy)
print(generator)
for gen in generator:
    print(gen)

[1, 100, 81, 0]
<generator object <genexpr> at 0x0000023938916660>
1
100
81
0


Advantages of Python Generators
Highly Memory Efficient
A normal function to return a sequence will create the entire sequence in memory before returning the result within the runtime itself.

3. Represent Infinite Stream
Generators can represent the length of a series of infinite numbers taken at a time. And this is supported by the library function.

4. Pipelining Generators in Python
Multiple generators can be used to pipeline a series of operations within the runtime itself.

```


In [8]:
#Cheecking memory utilization 
!pip install -U memory_profiler
!pip install Pympler

  Created wheel for Pympler: filename=Pympler-0.9-py3-none-any.whl size=164803 sha256=57530ebca76749acd34f222dce7299afdf32005201279127560d1723876e88e1
  Stored in directory: c:\users\dell\appdata\local\pip\cache\wheels\24\6f\0b\da9f81234859a8741aaea3afcc6ae2daf0efb67e7ff2d3686c
Successfully built Pympler


In [12]:
import memory_profiler as mem_profile
import random
import time
names = ['Sheetal ', 'Anupriya', 'Jasmine', 'Arya']
profession = ['Education', 'Engineer', 'Banker', 'Business']
print('Memory (before) : {}Mb'.format(mem_profile.memory_usage()))
t1 = time.time()
list = function_list(1000000)
t2 = time.time()

#t1 = time.time()
#generator = fun_generator(1000000)
#t2 = time.time()
print('Memory (After) : {}Mb'.format(mem_profile.memory_usage()))
print(' Time taken to create list  {} Seconds'.format(t2-t1))
 #output in case of list   
#Memory (before) : [490.8359375]Mb
#Memory (After) : [489.26171875]Mb
#Time taken to create list  2.301082000000001 Seconds


#output in case of generator
#

Memory (before) : [50.6171875]Mb
Memory (After) : [321.1484375]Mb
 Time taken to create list  1.152130126953125 Seconds


Output in case of Generator
Memory (before) : [50.6015625]Mb
Memory (After) : [50.6015625]Mb
 Time taken to create list  0.0 Seconds

output in case of list
Memory (before) : [50.6171875]Mb
Memory (After) : [321.1484375]Mb
 Time taken to create list  1.152130126953125 Seconds